core/modules/loader/DoListItem.h

```diff
@@ -0,0 +1,180 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ */
+#ifndef LSST_QSERV_LOADER_DOLISTITEM_H
+#define LSST_QSERV_LOADER_DOLISTITEM_H
+
+// system headers
+#include <chrono>
+#include <list>
+
+// Qserv headers
+#include "util/ThreadPool.h"
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+
+class Central;
+
+
+class TimeOut {
+public:
+    using TimePoint = std::chrono::system_clock::time_point;
+    using Clock = std::chrono::system_clock;
+
+    explicit TimeOut(std::chrono::milliseconds timeOut) : _timeOut(timeOut) {}
+
+    bool due() { return due(Clock::now()); }
+    bool due(TimePoint now) {
+        auto triggerDiff = std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+        return (triggerDiff > _timeOut);
+    }
+
+    void triggered() { return triggered(Clock::now()); }
+    void triggered(TimePoint now) {
+        _lastTrigger = now;
+    }
+
+    std::chrono::milliseconds timeLeft(TimePoint now) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut = timeOut; }
+    std::chrono::milliseconds getTimeOut() const { return _timeOut; }
+private:
+    // How much time since lastTrigger needs to pass before triggering.
+    std::chrono::milliseconds _timeOut; // default, 15 minutes
+    TimePoint _lastTrigger{std::chrono::seconds(0)};
+};
+
+
+/// Children of this class *MUST* be created with shared pointers.
+///
+/// A DoListItem is meant to checked periodically by the DoList
+/// at a low frequency (a couple of times a second to once every
+/// few hours or even days).
+/// The DoListItems can cycle forever by just remaining on the
+/// DoList where it will run their actions when the timer run out,
+/// which is useful for monitoring status.
+/// Or they can be setup to run until they have completed once,
+/// a oneShot, which is useful for looking up or inserting keys.
+///
+/// A typical action would be sending out a UDP request for status
+/// every few seconds until a response is received. Then, after a
+/// few minutes with no updates, repeating that request to make sure
+/// the status hasn't changed.
+/// The system is supposed to notify others on changes, but these
+/// notifications can lost, so it makes sense to ask for one if
+/// nothing has been received for a while.
+class DoListItem : public std::enable_shared_from_this<DoListItem> {
+public:
+    using Ptr = std::shared_ptr<DoListItem>;
+
+    /// Children of this class *MUST* be created with shared pointers.
+    /// A factory function to enforce this is not practical since
```

Comment:
I don't understand why you can't make the constructors protected and implement a factory function. Can you please explain the issue to me? (here or in person)

Reply:
This class can't be used standalone, so a factory function would never be used. The constructor can be made protected. There is already a pure virtual function but it would add another layer of protection.